src/google/protobuf/util/internal/utility.cc

```diff
@@ -302,7 +302,7 @@ bool IsMap(const google::protobuf::Field& field,
 string DoubleAsString(double value) {
   if (value == std::numeric_limits<double>::infinity()) return "Infinity";
   if (value == -std::numeric_limits<double>::infinity()) return "-Infinity";
-  if (::isnan(value)) return "NaN";
+  if (std::isnan(value)) return "NaN";
```

Comment:
Hello Feng Xiao,

you are right.  MathLimits IsNaN works fine.  I am made a silly mistake
earlier, which resulted in a compilation error.  It is corrected and
committed to GitHub.

Best regards,

Karol

On 29 June 2015 at 22:47, Feng Xiao <notifications@github.com> wrote:

> In src/google/protobuf/util/internal/utility.cc
> <https://github.com/google/protobuf/pull/551#discussion_r33511081>:
>
> > @@ -302,7 +302,7 @@ bool IsMap(const google::protobuf::Field& field,
> >  string DoubleAsString(double value) {
> >    if (value == std::numeric_limits<double>::infinity()) return "Infinity";
> >    if (value == -std::numeric_limits<double>::infinity()) return "-Infinity";
> > -  if (::isnan(value)) return "NaN";
> > +  if (std::isnan(value)) return "NaN";
>
> From reading this thread:
>
> http://stackoverflow.com/questions/570669/checking-if-a-double-or-float-is-nan-in-c
> what I get is that "value != value" may not work reliable.
>
> Could you elaborate why "MathLimits::IsNan(value)" doesn't work? To make
> the code portable we need a portable version of IsNan and I would prefer
> making the MathLimits one work for that purpose.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/protobuf/pull/551/files#r33511081>.
>